vertical-pod-autoscaler/e2e/updater.go

```diff
@@ -113,6 +181,106 @@ func setupHamsterDeployment(f *framework.Framework, cpu, memory string, replicas
+	hamsterContainer := setupHamsterContainer(cpu, memory)
+	rc := framework.RcByNameContainer("hamster-rc", 3, "k8s.gcr.io/ubuntu-slim:0.1",
+		hamsterLabels, hamsterContainer, nil)
+	if replicas != nil {
+		rc.Spec.Replicas = replicas
+	}
+	rc.Namespace = f.Namespace.Name
+	err := testutils.CreateRCWithRetries(f.ClientSet, f.Namespace.Name, rc)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	err = waitForRCPodsRunning(f, rc)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	pods, err := getHamsterPods(f)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	return rc, pods
+}
+
+func waitForRCPodsRunning(f *framework.Framework, rc *apiv1.ReplicationController) error {
+	return wait.PollImmediate(pollInterval, pollTimeout, func() (bool, error) {
+		podList, err := getHamsterPods(f)
+		if err != nil {
+			framework.Logf("Error listing pods, retrying: %v", err)
+			return false, nil
+		}
+		podsRunning := int32(0)
+		for _, pod := range podList.Items {
+			if pod.Status.Phase == apiv1.PodRunning {
+				podsRunning += 1
+			}
+		}
+		return podsRunning == *rc.Spec.Replicas, nil
+	})
+}
+
+func setupHamsterJob(f *framework.Framework, cpu, memory string, replicas *int32) (*batchv1.Job, *apiv1.PodList) {
+	parallelism := int32(3)
+	if replicas != nil {
+		parallelism = *replicas
+	}
+	job := framework.NewTestJob("notTerminate", "hamster-job", apiv1.RestartPolicyOnFailure,
+		parallelism, parallelism, nil, 10)
+	job.Spec.Template.Spec.Containers[0] = setupHamsterContainer(cpu, memory)
+	job.Spec.Template.Labels["app"] = "hamster"
+	err := testutils.CreateJobWithRetries(f.ClientSet, f.Namespace.Name, job)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	err = framework.WaitForAllJobPodsRunning(f.ClientSet, f.Namespace.Name, job.Name, parallelism)
+	podList, err := framework.GetJobPods(f.ClientSet, f.Namespace.Name, job.Name)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	return job, podList
+}
+
+func setupHamsterRS(f *framework.Framework, cpu, memory string, replicas *int32) (*appsv1.ReplicaSet, *apiv1.PodList) {
+	rs := framework.NewReplicaSet("hamster-rs", f.Namespace.Name, 3,
+		hamsterLabels, "", "")
+	if replicas != nil {
+		rs.Spec.Replicas = replicas
+	}
+	rs.Spec.Template.Spec.Containers[0] = setupHamsterContainer(cpu, memory)
+	err := createReplicaSetWithRetries(f.ClientSet, f.Namespace.Name, rs)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	err = framework.WaitForReadyReplicaSet(f.ClientSet, f.Namespace.Name, rs.Name)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	podList, err := getHamsterPods(f)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	return rs, podList
+}
+
+func setupHamsterStateful(f *framework.Framework, cpu, memory string, replicas *int32) (*appsv1.StatefulSet,
+	*framework.StatefulSetTester, *apiv1.PodList) {
+	stateful := framework.NewStatefulSet("hamster-stateful", f.Namespace.Name,
+		"hamster-service", 3, nil, nil, hamsterLabels)
```

Comment:
Same comment as with setupHamsterRS.

And, btw, could you please extract 3 to a const? We have it in all setupHamsterXYZ methods and it always means "default number of replicas".
Even better maybe we could require this number to be always set and just populate it from testRestartsPods or setupHamsterController. This would simplify the awkward call in "observes pod disruption budget" test too.

Reply:
Done